### ISSUE TITLE
Failed to mount nfs share : mount_cb: RPC error: Mount failed with error MNT3ERR_ACCES(13) Permission denied(13)

### ISSUE THREAD

**Issue by nufeng1999** [sahlberg/fuse-nfs#43]
Failed to mount nfs share : mount_cb: RPC error: Mount failed with error MNT3ERR_ACCES(13) Permission denied(13)

rpcinfo -p 192.168.0.98
   program vers proto   port  service
    100000    4   tcp    111  portmapper
    100000    3   tcp    111  portmapper
    100000    2   tcp    111  portmapper
    100000    4   udp    111  portmapper
    100000    3   udp    111  portmapper
    100000    2   udp    111  portmapper
    100005    1   udp  20048  mountd
    100005    1   tcp  20048  mountd
    100005    2   udp  20048  mountd
    100005    2   tcp  20048  mountd
    100005    3   udp  20048  mountd
    100005    3   tcp  20048  mountd
    100003    2   tcp   2049  nfs
    100003    3   tcp   2049  nfs
    100003    4   tcp   2049  nfs
    100021    1   tcp  39647  nlockmgr
    100021    3   tcp  39647  nlockmgr
    100021    4   tcp  39647  nlockmgr
    100003    2   udp   2049  nfs
    100003    3   udp   2049  nfs
    100003    4   udp   2049  nfs
    100021    1   udp  56298  nlockmgr
    100021    3   udp  56298  nlockmgr
    100021    4   udp  56298  nlockmgr
fuse-nfs -n nfs://192.168.0.98/media/usbhd-sdb1 -m /mnt/work1
Failed to mount nfs share : mount_cb: RPC error: Mount failed with error MNT3ERR_ACCES(13) Permission denied(13)

 cat exports
/media/usbhd-sdb1 *(rw,no_root_squash,sync,insecure)

**Comment by pkthapa** [comment]
Run below commands on NFS server after adding `/media/usbhd-sdb1 *(rw,no_root_squash,sync,insecure)` in `/etc/exports` of NFS server.
```
sudo exportfs -a #making the file share available 
sudo systemctl restart nfs-kernel-server #restarting the NFS kernel
```
However, I want to use server's credentials, from the `fuse-nfs` client for mounting and unmounting, rather than modifying `/etc/exports` of the server. Not sure how to achieve this.
@sahlberg @nufeng1999

**Comment by webghostx** [comment]
Indeed, the "insecure" option must be set. Then I found out that the directory for the mount point must not be write-protected.

This seems to me to cause two problems:
1) The security aspect regarding ports is no longer relevant.
2) I cannot use write protection to prevent writing if the mount is not present.

I see the second point as problematic, or how can I solve it?

```
yves@yoga:~$ fuse-nfs -n nfs://10.0.0.14/export/Public -m /m/naspi/Public
Failed to mount nfs share : mount_cb: RPC error: Mount failed with error MNT3ERR_ACCES(13) Permission denied(13)
# add insecure to nfs options
yves@yoga:~$ fuse-nfs -n nfs://10.0.0.14/export/Public -m /m/naspi/Public
fusermount: failed to chdir to mountpoint: Permission denied
# change chmod from 444 to 744
yves@yoga:~$ fuse-nfs -n nfs://10.0.0.14/export/Public -m /m/naspi/Public
yves@yoga:~$ 
```

**Comment by pkthapa** [comment]
Thank you @webghostx for your response.
On the similar lines, I have another query. Is there a way to prevent mounting of NFS share using `UID`, `GID`, `--fusenfs_uid`, or `--fusenfs_gid` option?
I tried this command and was able to mount successfully: `fuse-nfs --fusenfs_uid=7000 --fusenfs_gid=8000 -n nfs://<nfs-server-ip>/mnt/pkt -m /home/pankaj/pkt`, however my expectation was mount failure. The NFS server mount path [`/mnt/pkt`] had different `UID` & `GID`.
I appreciate if you can throw some light here!!! Thank you once again.
cc: @sahlberg @nufeng1999

**Comment by pkthapa** [comment]
Also, I would like to know one thing:-
My nfs client (Ubuntu) has firewall enabled. I usually run below commands on the client machine to allow (bypass firewall) all ports for `fuse-nfs` mounting.
```
iptables -P INPUT ACCEPT
iptables -P OUTPUT ACCEPT
iptables -P FORWARD ACCEPT
```
However, I don't want to run these 3 `iptables` command. Rather, I tried allowing **udp** and **tcp** both on these ports (`111`, `2049`, `32765`, `32766`, `32767`) but failed to mount again. Am I missing something here? Can anyone point me out my mistake?
Thank you once again.

`Note: Both NFS server and client are running on Ubuntu OS`.

cc: @webghostx @nufeng1999 @sahlberg

**Comment by sahlberg** [comment]
NFSv3 only uses fixed ports for NFS (2049) and portmapper(111) but ephemeral ports for all other services.

Your distro should have scripts to dynamically upen ports for things like MOUNT and LOCKD

Otherwise, switch to using NFSv4. It uses only port 2049, so open tcp 2049 and make sure to use NFSv4
